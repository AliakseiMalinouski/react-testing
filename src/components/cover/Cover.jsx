import React from "react";
import { useState } from "react";

const Cover = ({ title, text, items }) => {
    const [data, setData] = useState(items);
    const handleClick = (id) => {
        const preparedData = data.map((item) => {
            return {
                ...item,
                checked: item.id === id,
            }
        });
        setData(preparedData);
    }
    return (
        <div>
            <h1>{title}</h1>
            <p>{text}</p>
            <div data-testid="items-test-id">
            {data.map(({ title, checked, id }) => (
                <div key={id} onClick={() => handleClick(id)} data-testid={`item-test-id-${id}`}>
                    <h3>{title}</h3>
                    <input type='checkbox' checked={checked}/>
                </div>
            ))}
            </div>
        </div>
    )
}

export {
    Cover,
}
