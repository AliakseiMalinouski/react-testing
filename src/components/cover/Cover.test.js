import React from "react";
import { Cover } from "./Cover";
import { render, fireEvent } from "@testing-library/react";

const text = 'some text';
const title = 'some title';
const items = [
    {
        id: 1,
        title: 'item1',
        checked: false,
    },
    {
        id: 2,
        title: 'item2',
        checked: false,
    },
    {
        id: 3,
        title: 'item3',
        checked: false,
    }
]

describe('Cover tests', () => {
    test('Cover render test', () => {
       const { container } = render(<Cover text={text} title={title} items={items}/>);
       expect(container).toMatchSnapshot();
    });

    test('Should display title', () => {
        const { queryByText } = render(<Cover text={text} title={title} items={items}/>);
        const titleQuery = queryByText(title);
        expect(titleQuery).toBeTruthy();
    });

    test('Should display text', () => {
        const { queryByText } = render(<Cover text={text} title={title} items={items}/>);
        const textQuery = queryByText(text);
        expect(textQuery).toBeTruthy();
    });

    test('Should display items', () => {
        const { getByTestId } = render(<Cover text={text} title={title} items={items}/>);
        const minItemsLength = 2;
        const testiId = 'items-test-id';
        const itemsWrapper = getByTestId(testiId);
        expect(itemsWrapper.children.length).toBeGreaterThan(minItemsLength);
    });

    test('onClick test', () => {
        const { getByTestId } = render(<Cover text={text} title={title} items={items}/>);
        const testiId = 'items-test-id';
        const { id } = Object.values(items)[Math.floor(Math.random()) * items.length];
        const randomItemId = `item-test-id-${id}`;
        const randomItem = getByTestId(randomItemId);
        fireEvent.click(randomItem);
    });

    test('onClick test trigger', () => {
        const onClickTrigger = jest.fn();
        const { container } = render(<Cover text={text} title={title} items={items} onClick={onClickTrigger}/>);
    });

    test('item button test', () => {
        const onClickTrigger = jest.fn();
        const { container, getByRole } = render(<Cover text={text} title={title} items={items} onClick={onClickTrigger}/>);
        expect(getByRole('button', { type: 'button' })).toBeTruthy();
    });

    test('input value test', () => {
        const onClickTrigger = jest.fn();
        const inputValue = 'input value';
        const { container, getByRole, getByTestId } = render(<Cover text={text} title={title} items={items} inputValue={inputValue} onClick={onClickTrigger}/>);
        expect(getByTestId("input-with-prop-value").value).toBe(inputValue);
    });
});