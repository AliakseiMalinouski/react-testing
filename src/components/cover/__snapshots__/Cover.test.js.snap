// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cover tests Cover render test 1`] = `
<div>
  <div>
    <h1>
      some title
    </h1>
    <p>
      some text
    </p>
    <div
      data-testid="items-test-id"
    >
      <div
        data-testid="item-test-id-1"
      >
        <h3>
          item1
        </h3>
        <input
          type="checkbox"
        />
        <button
          type="button"
        />
      </div>
      <div
        data-testid="item-test-id-2"
      >
        <h3>
          item2
        </h3>
        <input
          type="checkbox"
        />
        <button
          type="button"
        />
      </div>
      <div
        data-testid="item-test-id-3"
      >
        <h3>
          item3
        </h3>
        <input
          type="checkbox"
        />
        <button
          type="button"
        />
      </div>
    </div>
  </div>
</div>
`;
